protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,798 @@
+    auto_retry: Option<(AutoRetry, Delay)>,
+
+    // Out events that should be reported to the user
+    pending_out_event: VecDeque<NatStatus>,
+
+    next_probe_id: ProbeId,
+}
+
+impl Behaviour {
+    pub fn new(config: Config) -> Self {
+        let proto_support = match config.server {
+            Some(_) => ProtocolSupport::Full,
+            None => ProtocolSupport::Outbound,
+        };
+        let protocols = iter::once((AutoNatProtocol, proto_support));
+        let mut cfg = RequestResponseConfig::default();
+        cfg.set_request_timeout(config.timeout);
+        let inner = RequestResponse::new(AutoNatCodec, protocols, cfg);
+        let auto_retry = config.auto_retry.map(|a| (a, Delay::new(Duration::ZERO)));
+        Self {
+            inner,
+            ongoing_inbound: HashMap::default(),
+            ongoing_outbound: None,
+            pending_probe: None,
+            connected: HashMap::default(),
+            reachability: Reachability::Unknown,
+            server_config: config.server,
+            auto_retry,
+            pending_out_event: VecDeque::new(),
+            next_probe_id: ProbeId(1),
+        }
+    }
+
+    // Manually retry determination of NAT status.
+    pub fn retry_nat_status(&mut self, probe_config: ProbeConfig) -> ProbeId {
+        let id = self.next_probe_id();
+        self.pending_probe.replace((id, probe_config));
+        id
+    }
+
+    // Assumed reachability derived from the most recent probe.
+    pub fn reachability(&self) -> Reachability {
+        self.reachability.clone()
+    }
+
+    /// Add peer to the list of trusted public peers that are probed in the auto-retry nat status.
+    ///
+    /// Return false if auto-retry is `None`.
+    pub fn add_server(&mut self, peer: PeerId, address: Option<Multiaddr>) -> bool {
+        match self.auto_retry {
+            Some((ref mut retry, _)) => {
+                retry.config.servers.push(peer);
+                if let Some(addr) = address {
+                    self.inner.add_address(&peer, addr);
+                }
+                true
+            }
+            None => false,
+        }
+    }
+
+    /// Remove a peer from the list of servers tried in the auto-retry.
+    ///
+    /// Return false if auto-retry is `None`.
+    pub fn remove_server(&mut self, peer: &PeerId) -> bool {
+        match self.auto_retry {
+            Some((ref mut retry, _)) => {
+                retry.config.servers.retain(|p| p != peer);
+                true
+            }
+            None => false,
+        }
+    }
+
+    fn next_probe_id(&mut self) -> ProbeId {
+        let probe_id = self.next_probe_id;
+        self.next_probe_id.0 += 1;
+        probe_id
+    }
+
+    fn do_probe(&mut self, mut probe: Probe) {
+        for (peer_id, id) in probe.pending_servers.iter_mut() {
+            let request_id = self.inner.send_request(
+                peer_id,
+                DialRequest {
+                    peer_id: *peer_id,
```

Comment:
This should be the local peer ID and not the peer ID of some probing server, right?

Go servers return a `BadRequest` `peer id mismatch` to me.

https://github.com/libp2p/go-libp2p-autonat/blob/8ff057fa9a70c29be0a8eaf29c98a37d5e732767/svc.go#L97-L99

Changing this to the local peer ID solves the problem.

Reply:
Afaik there is no way from within a `NetworkBehaviour` to get the own peer-id, so it now has to be entered in `Behaviour::new`. Not the best solution because the user has to enter it twice when creating the Swarm (since the swarm itself also needs it), but I don't think there's any other way to solve this.

Reply:
>  Afaik there is no way from within a NetworkBehaviour to get the own peer-id, so it now has to be entered in Behaviour::new. Not the best solution because the user has to enter it twice when creating the Swarm (since the swarm itself also needs it), but I don't think there's any other way to solve this.

Agreed that having to pass the peer ID twice, once to `Swarm`, once to `AutoNAT` is silly. I have to do the same in the circuit relay v2 implementation https://github.com/libp2p/rust-libp2p/pull/2059. Unfortunately there is no other way today. Also I fail to come up with a good mechanism providing the local peer ID from the `Swarm`. Happy for suggestions.

>  Kinda related to what you wrote in libp2p/specs#362 (comment); do we even need the peer-id here at all? But I guess it would break go-autonat to remove it from the structure, so it think the current solution is fine.

Yes, I don't understand why we send it. I hope one day we will fix this in AutoNAT v2.